src/Build/BackEnd/Components/Communications/NodeProviderOutOfProcTaskHost.cs

```diff
@@ -474,20 +471,67 @@ internal static string GetMSBuildExecutablePathForNonNETRuntimes(HandshakeOption
         {
             ErrorUtilities.VerifyThrowInternalErrorUnreachable(Handshake.IsHandshakeOptionEnabled(hostContext, HandshakeOptions.TaskHost));
 
-            s_baseTaskHostPathNet ??= BuildEnvironmentHelper.Instance.MSBuildToolsDirectoryNET;
+            taskHostParameters.TryGetValue(Constants.DotnetHostPath, out string runtimeHostPath);
+            var msbuildAssemblyPath = GetMSBuildAssemblyPath(taskHostParameters);
 
-            string msbuildAssemblyPath = Path.Combine(BuildEnvironmentHelper.Instance.MSBuildAssemblyDirectory, Constants.MSBuildAssemblyName);
-            string runtimeHostPath = Path.Combine(s_baseTaskHostPathNet, Constants.DotnetProcessName);
-
-            // Current process is .NET Framework, but handshake requests .NET
-            // Launch dotnet host and path to MSBuild.dll
             return (runtimeHostPath, msbuildAssemblyPath);
         }
 
+        private static string GetMSBuildAssemblyPath(Dictionary<string, string> taskHostParameters)
+        {
+            if (taskHostParameters.TryGetValue(Constants.MSBuildAssemblyPath, out string msbuildAssemblyPath))
+            {
+                ValidateNetHostSdkVersion(msbuildAssemblyPath);
+
+                return msbuildAssemblyPath;
+            }
+
+            throw new InvalidProjectFileException(ResourceUtilities.GetResourceString("NETHostTaskLoad_Failed"));
+
+            static void ValidateNetHostSdkVersion(string path)
+            {
+                const int MinimumSdkVersion = 10;
+
+                if (string.IsNullOrEmpty(path))
+                {
+                    ErrorUtilities.ThrowInternalError(ResourceUtilities.GetResourceString("SDKPathResolution_Failed"));
+                }
+
+                if (!FileSystems.Default.DirectoryExists(path))
+                {
+                    ErrorUtilities.ThrowInternalError(ResourceUtilities.FormatResourceStringIgnoreCodeAndKeyword("SDKPathCheck_Failed", path));
+                }
+
+                var sdkVersion = ExtractSdkVersionFromPath(path);
+                if (sdkVersion is null or < MinimumSdkVersion)
+                {
+                    throw new InvalidProjectFileException(ResourceUtilities.FormatResourceStringIgnoreCodeAndKeyword("NETHostVersion_Failed", sdkVersion, MinimumSdkVersion));
+                }
+            }
+        }
+
+        private static int? ExtractSdkVersionFromPath(string path)
+        {
+            string lastDirectoryName = Path.GetFileName(path.TrimEnd(Path.DirectorySeparatorChar));
+
+            if (string.IsNullOrEmpty(lastDirectoryName))
+            {
+                return null;
+            }
+
+            int dotIndex = lastDirectoryName.IndexOf('.');
+            if (dotIndex <= 0)
+            {
+                return null;
+            }
+
+            return int.TryParse(lastDirectoryName.Substring(0, dotIndex), out int majorVersion) ? majorVersion : 0;
```

Comment:
What's the `0` mean here? Is that not another `null` case?

Reply:
0 is a version that is below minimum , I will change it to null for consistency.